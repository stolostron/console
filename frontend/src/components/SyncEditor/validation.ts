/* Copyright Contributors to the Open Cluster Management project */
import Ajv from 'ajv'
import stringSimilarity from 'string-similarity'
import { isEmpty, get, set, keyBy, cloneDeep } from 'lodash'
import { getPathArray } from './synchronize'
import { MappingType } from './process'

export enum ErrorType {
  error = 'error',
  warning = 'warning',
  info = 'info',
}

export interface ErrorMessageType {
  linePos: {
    end: { line: number; col: number }
    start: { line: number; col: number }
  }
  message: string
  errorType?: ErrorType
}

//////// compile ajv schemas for validation
export const compileAjvSchemas = (schema: any[]) => {
  try {
    const ajv = new Ajv({ allErrors: true, verbose: true })
    addAjvKeywords(ajv)
    if (!Array.isArray(schema)) {
      return [{ validator: ajv.compile(schema) }]
    } else {
      const schemas: any = []
      schema.forEach(({ type, required, schema }) => {
        schemas.push({
          type,
          required,
          validator: ajv.compile(schema),
        })
      })
      return schemas
    }
  } catch (e) {
    console.log(e)
  }
}

export const addAjvKeywords = (ajv: Ajv) => {
  ajv.addKeyword({
    keyword: 'validateName',
    schemaType: 'boolean',
    validate: (_schema: null, data: any) => {
      return (
        !data || (/^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$/.test(data) && data.length <= 253)
      )
    },
  })
  ajv.addKeyword({
    keyword: 'validateTemplateName',
    schemaType: 'boolean',
    validate: (_schema: null, data: any) => {
      return (
        !data ||
        (typeof data === 'string' &&
          (data.startsWith('{')
            ? /^(\{\{)[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*(\}\})$/.test(data)
            : /^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$/.test(data)) &&
          data.length <= 63)
      )
    },
  })
  ajv.addKeyword({
    keyword: 'validateDep',
    schemaType: 'boolean',
    validate: (_schema: null, data: any) => {
      if (data['namespace']) {
        if (
          (data['kind'] && data['kind'] === 'ConfigurationPolicy') ||
          data['kind'] == 'IamPolicy' ||
          data['kind'] == 'CertificatePolicy'
        ) {
          return false
        }
      }
      return true
    },
  })
  ajv.addKeyword({
    keyword: 'validateLabel',
    schemaType: 'boolean',
    validate: (_schema: null, data: any) => {
      return (
        !data || (/^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$/.test(data) && data.length <= 63)
      )
    },
  })
  ajv.addKeyword({
    keyword: 'deprecatedKind',
    schemaType: 'string',
    validate: () => {
      return false
    },
  })
}

//////// syntax errors
export const getErrors = (documents: any[]) => {
  const errors: any[] = []
  documents.forEach((document: { errors: any[] }) => {
    document?.errors.forEach((error: { linePos: any; message: any }) => {
      const { linePos, message } = error
      errors.push({ linePos, message })
    })
  })
  return errors
}

///////// validation errors
export function validate(
  validators: any,
  mappings: { [name: string]: any[] },
  resources: any[],
  errors: any[],
  syntaxErrors: any[],
  protectedRanges: any[]
) {
  const kindMap = {}
  const validatorMap = keyBy(validators, 'type')
  const requiredTypes: Record<string, number> = {}
  Object.entries(validatorMap).forEach(([k, v]) => {
    if (v.required) {
      requiredTypes[k] = v.required
    }
  })
  const wasRequired = cloneDeep(requiredTypes)
  resources.forEach((resource) => {
    const kind = resource.kind
    requiredTypes[kind]--
    let d: number = get(kindMap, kind, 0)
    // determine validation for this resource
    let validator = validators[0]?.validator
    if (validators.length > 1) {
      validator = validatorMap[kind]?.validator
      if (!validator && kind) {
        const matches = stringSimilarity.findBestMatch(kind, Object.keys(validatorMap))
        const {
          bestMatch: { rating, target },
        } = matches
        if (rating > 0.7) {
          validator = validatorMap[target]?.validator
        }
      }
    }
    if (validator) {
      // d keeps track of Secret[0], Secret[1], Secret[d]
      validateResource(validator, mappings, [resource.kind, d], resource, errors, protectedRanges)
      set(kindMap, resource.kind, d++)
    }
  })
  // what required types weren't found
  Object.entries(requiredTypes).forEach(([k, v]) => {
    if (v > 0) {
      syntaxErrors.push({
        linePos: {
          end: { line: 1, col: 1 },
          start: { line: 1, col: 1 },
        },
        message: `Requires ${wasRequired[k]} ${k}`,
      })
    }
  })
}

export function validateResource(
  validator: any,
  mappings: { [name: string]: any[] },
  prefix: string[],
  resource: any,
  errors: any[],
  protectedRanges: any[]
) {
  const valid = validator(resource)
  if (!valid) {
    validator.errors.forEach((error: { instancePath: string; keyword: string; message: string; params: any }) => {
      const { instancePath, keyword, message, params } = error
      const errorMsg: ErrorMessageType = {
        linePos: {
          end: { line: 1, col: 1 },
          start: { line: 1, col: 1 },
        },
        message,
      }

      // convert instance path to line
      // for paths, lodash get hates periods, ajv hates forward slash
      let path: string[] = prefix
      if (!isEmpty(instancePath)) {
        path = instancePath
          .substring(1)
          .split('/')
          .map((p) => {
            return p.replace(/~\d+/g, '/')
          })
        path = [...prefix, ...path]
        path = getPathArray(path)
      }
      let mapping = get(mappings, path)
      if (mapping) {
        // no need to show validation errors on immutable lines
        const lineNumber = mapping?.kind?.$r ?? mapping?.$r ?? 1
        const lineIsProtected = protectedRanges.some((prohibited: { containsPosition: (arg: any) => any }) => {
          return prohibited.containsPosition({ lineNumber, startColumn: 1, endColumn: 1 })
        })
        if (!lineIsProtected) {
          errorMsg.linePos.start.line = errorMsg.linePos.end.line = lineNumber
          errorMsg.errorType = ErrorType.error
          let matches
          switch (keyword) {
            // missing a key
            case 'required':
              // see if there's a misspelled key
              mapping = mapping.$v || mapping
              if (!isEmpty(mapping)) {
                matches = stringSimilarity.findBestMatch(params.missingProperty, Object.keys(mapping))
              }
              if (matches) {
                const {
                  bestMatch: { rating, target },
                } = matches
                if (rating > 0.7) {
                  const similar = mapping[target]
                  errorMsg.linePos.start.line = errorMsg.linePos.end.line = similar.$r
                  errorMsg.linePos.start.col = similar.$gk.start.col
                  errorMsg.linePos.end.col = similar.$gk.end.col
                }
              }
              break
            case 'const':
              errorMsg.message = `${message}: ${params.allowedValue}`
              errorMsg.linePos.start.col = mapping.$gv.start.col
              errorMsg.linePos.end.col = mapping.$gv.end.col
              break
            case 'deprecatedKind':
              errorMsg.message = `This kind is deprecated: ${mapping.$v}`
              errorMsg.linePos.start.col = mapping.$gv.start.col
              errorMsg.linePos.end.col = mapping.$gv.end.col
              errorMsg.errorType = ErrorType.info
              break
            // value wrong pattern
            case 'pattern':
              errorMsg.linePos.start.col = mapping.$gv.start.col
              errorMsg.linePos.end.col = mapping.$gv.end.col
              break
            // validateName
            case 'validateName':
              errorMsg.message =
                'Name must start/end alphanumerically, can contain dashes and periods, and must be less then 253 characters'
              errorMsg.linePos.start.col = mapping.$gv.start.col
              errorMsg.linePos.end.col = mapping.$gv.end.col
              break
            // validateDep
            case 'validateDep':
              errorMsg.message =
                'Dependencies on ConfigurationPolicies, IamPolicies, and CertificatePolicies cannot contain a namespace'
              errorMsg.linePos.start.line = mapping.$gv.start.line
              errorMsg.linePos.end.line = mapping.$gv.end.line
              errorMsg.linePos.start.col = mapping.$gv.start.col
              errorMsg.linePos.end.col = mapping.$gv.end.col
              break
            // validateLabel
            case 'validateLabel':
              errorMsg.message =
                'Name must start/end alphanumerically, can contain dashes, and must be less then 63 characters'
              errorMsg.linePos.start.col = mapping.$gv.start.col
              errorMsg.linePos.end.col = mapping.$gv.end.col
              break
            // value wrong enum
            case 'enum':
              errorMsg.message = `${message}: ${params.allowedValues
                .map((val: any) => {
                  return `"${val}"`
                })
                .join(', ')}`
              if (mapping.$gv) {
                errorMsg.linePos.start.col = mapping.$gv.start.col
                errorMsg.linePos.end.col = mapping.$gv.end.col
              }
              errorMsg.errorType = ErrorType.warning
              break
            case 'type':
              errorMsg.message = message
              errorMsg.linePos.start.col = mapping.$gv?.start?.col ?? 1
              errorMsg.linePos.end.col = mapping.$gv?.end?.col ?? 1
              break
            default:
              errorMsg.message = message
              errorMsg.linePos.start.col = mapping.$gv?.start?.col ?? 1
              errorMsg.linePos.end.col = mapping.$gv?.end?.col ?? 1
              break
          }
          errors.push(errorMsg)
        }
      }
    })
  }
  validateTemplateSyntax(mappings, errors)
}

export const validateTemplateSyntax = (object: any, errors: any[]) => {
  if (object) {
    object = (object.$v !== undefined ? object.$v : object) as { [name: string]: any | any[] }
    if (Array.isArray(object)) {
      object.forEach((o) => {
        validateTemplateSyntax(o, errors)
      })
    } else if (!!object && typeof object === 'object') {
      Object.values(object).forEach((o) => {
        const obj = o as unknown as MappingType
        if (
          obj &&
          Array.isArray(obj.$p) &&
          ['metadata.name', 'metadata.namespace'].includes(obj.$p.slice(-2).join('.')) &&
          typeof obj.$v !== 'string'
        ) {
          /* istanbul ignore next */
          const errorMsg: ErrorMessageType = {
            linePos: {
              end: { line: obj.$r, col: obj.$gv ? obj.$gv.end.col : 1 },
              start: { line: obj.$r, col: obj.$gv ? obj.$gv.start.col : 1 },
            },
            errorType: ErrorType.error,
            message: 'If this is a template name, it must be in the form "{{ name }}"',
          }
          errors.push(errorMsg)
        } else {
          validateTemplateSyntax(o, errors)
        }
      })
    }
  }
}

export const formatErrors = (errors: ErrorMessageType[], merrorType?: ErrorType) => {
  return errors
    .filter(({ errorType }) => errorType === merrorType)
    .map((error) => {
      return {
        line: error.linePos.start.line,
        col: error.linePos.start.col,
        message: error.message,
      }
    })
}
