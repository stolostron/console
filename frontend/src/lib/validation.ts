/* Copyright Contributors to the Open Cluster Management project */

import YAML from 'yaml'

import { TFunction } from 'i18next'
import validator from 'validator'
import { IResource } from '../resources'

import isCidr from 'is-cidr'

const lowercaseAlphaNumericCharacters = 'abcdefghijklmnopqrstuvwxyz1234567890'

export function validateKubernetesDnsName(value: string, t: TFunction) {
  if (value) {
    if (value.length > 63) return `${t('validate.kubernetesDnsName.length')}`
    for (const char of value) {
      if (!lowercaseAlphaNumericCharacters.includes(char) && char !== '-')
        return `${t('validate.kubernetesDnsName.char')}`
    }
    if (!lowercaseAlphaNumericCharacters.includes(value[0])) return `${t('validate.kubernetesDnsName.startchar')}`
    if (!lowercaseAlphaNumericCharacters.includes(value[value.length - 1]))
      return `${t('validate.kubernetesDnsName.endchar')}`
  }
  return undefined
}

export function validatePublicSshKey(value: string, t: TFunction, required: boolean) {
  if (!value && !required) {
    return undefined
  }
  if (value) {
    // Public SSH key should start with 'ssh-rsa' or 'ssh-dss', for example
    // Second token is a base64 value, with first integer being the length of the first token
    // (eg. 7 for RSA and DSA keys, 11 for ed25519 keys, etc.)
    const keyTypes = ['ssh-rsa', 'ssh-dss', 'ssh-ed25519', 'ecdsa-sha2-nistp256', 'ecdsa-sha2-nistp521']
    const tokens = value.trim().split(/\s+/)
    if (tokens.length >= 2) {
      if (keyTypes.includes(tokens[0])) {
        try {
          const firstInteger = Buffer.from(tokens[1], 'base64').readInt32BE(0)
          if (firstInteger === tokens[0].length) {
            // Valid key; exit validation
            return undefined
          }
        } catch (e) {
          // Fall through to error case
        }
      }
    }
  }
  return t('validate.publicSshKey')
}

export function validatePrivateSshKey(value: string, t: TFunction, requireNewline = true) {
  if (!/-----BEGIN [a-zA-Z]+ PRIVATE KEY-----\n([\s\S]*?)\n-----END [a-zA-Z]+ PRIVATE KEY-----/gm.test(value)) {
    return t('validate.privateSshKey')
  }
  if (requireNewline && !/[\r\n]$/.test(value)) {
    return t('validate.mustEndWithNewline')
  }
}

export function validateCertificate(value: string, t: TFunction) {
  if (!/-----BEGIN CERTIFICATE-----\n([\s\S]*?)\n-----END CERTIFICATE-----/gm.test(value)) {
    return t('validate.certificate')
  }
  return undefined
}

export function validateGCProjectID(value: string, t: TFunction) {
  const projectIDPattern = /^[a-z][a-z0-9-]{4,28}[a-z0-9]$/
  if (!projectIDPattern.test(value)) {
    return t('validate.projectID.format')
  }

  return undefined // the value is valid
}

export function validateJSON(value: string, t: TFunction) {
  try {
    const obj = JSON.parse(value)
    if (Object.entries(obj).length <= 0) {
      return t('validate.json')
    }
  } catch (e) {
    return t('validate.json')
  }
  return undefined
}

export const VALID_DNS_NAME_TESTER = new RegExp('^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$')
export function validateBaseDnsName(value: string, t: TFunction) {
  if (value && value.startsWith('.') && VALID_DNS_NAME_TESTER.test(value.substr(1))) {
    return t('validate.baseDnsName.start')
  }
  if (!VALID_DNS_NAME_TESTER.test(value)) {
    return t('validate.baseDnsName.char')
  }

  return undefined
}

export function validateLibvirtURI(value: string, t: TFunction) {
  const VALID_LIBVIRT_PROTOCOLS = ['qemu+ssh']
  const protoValuePair = value.split('://')
  if (protoValuePair.length !== 2 || !VALID_LIBVIRT_PROTOCOLS.includes(protoValuePair[0])) {
    return t('validate.libevirtURI.format')
  }

  if (!protoValuePair[1]) {
    // whatever but not empty
    return t('validate.libevirtURI.format')
  }

  return undefined // the value is valid
}

export function validateBaseDomain(value: string, t: TFunction) {
  const VALID_DNS_NAME_TESTER = /^[a-z0-9]([-a-z0-9]*[a-z0-9])?(.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$/
  if (value) {
    if (value.startsWith('.') && VALID_DNS_NAME_TESTER.test(value.substr(1))) {
      return t('validate.baseDomain.baseDNSPeriod')
    }
    if (!VALID_DNS_NAME_TESTER.test(value)) {
      return t('validate.baseDomain.name')
    }
  }
  return undefined
}

export function validateCloudsYaml(yamlValue: string, cloudValue: string, t: TFunction) {
  if (yamlValue) {
    try {
      //ensure we have valid YAML
      const yamlData = YAML.parse(yamlValue) as {
        clouds: {
          [cloud: string]: {
            auth?: {
              auth_url?: string
              password?: string
              username?: string
            }
          }
        }
      }

      //check for the clouds key
      const clouds = yamlData.clouds
      if (clouds === undefined) {
        return t('validate.yaml.not.valid')
      }

      let found = false
      for (const key in clouds) {
        //look for matching cloud name
        if (cloudValue !== undefined && key === cloudValue) {
          found = true
        }
        //check a few of the required fields, especially password, since the user
        //would have had to add this manually
        if (
          clouds[key]?.auth?.auth_url === undefined ||
          clouds[key]?.auth?.password === undefined ||
          clouds[key]?.auth?.username === undefined
        ) {
          return t('validate.yaml.not.valid')
        }
      }
      //Uh-oh, cloud name not found in clouds.yaml
      if (cloudValue !== undefined && !found) {
        return t('validate.yaml.cloud.not.found')
      }
    } catch (e) {
      return t('validate.yaml.not.valid')
    }
  }
  return undefined
}

export function validateAnsibleHost(url: string, t: TFunction, supportedProtocols?: string[]) {
  const protocols = supportedProtocols ? supportedProtocols : ['http', 'https']
  if (
    validator.isURL(url, {
      require_protocol: true,
      require_valid_protocol: true,
      protocols: protocols,
      require_host: true,
    })
  )
    return undefined
  return t('validate.ansible.url.not.valid')
}

export function validateWebURL(url: string, _item: unknown, t?: TFunction) {
  t = t ? t : (value) => value
  if (
    validator.isURL(url, {
      require_protocol: true,
      require_valid_protocol: true,
      protocols: ['http', 'https'],
      require_host: true,
    })
  )
    return undefined
  return `${t('The URL is not valid.')}`
}

export function validateImageContentSources(value: string, t: TFunction) {
  if (value) {
    try {
      //ensure we have valid YAML
      const yamlData: { mirrors?: string[]; source?: string }[] = YAML.parse(value)
      const isValid = yamlData.every((item) => {
        return Array.isArray(item.mirrors) && item.source
      })

      //check for the clouds key
      if (!isValid) {
        return t('validate.yaml.not.valid')
      }
    } catch (e) {
      return t('validate.yaml.not.valid')
    }
  }
  return undefined
}

export function validateYAML(value: string, t: TFunction) {
  if (value) {
    try {
      YAML.parse(value)
    } catch (e) {
      return t('validate.yaml.not.valid')
    }
  }
  return undefined
}

export function validateHttpProxy(value: string, t: TFunction) {
  if (value) {
    if (
      validator.isURL(value, {
        require_protocol: true,
        require_valid_protocol: true,
        protocols: ['http'],
        require_host: true,
      })
    )
      return undefined
    return t('validate.http.proxy.url.not.valid')
  }
  return undefined
}

export function validateHttpsProxy(value: string, t: TFunction) {
  if (value) {
    if (
      validator.isURL(value, {
        require_protocol: true,
        require_valid_protocol: true,
        protocols: ['http', 'https'],
        require_host: true,
      })
    )
      return undefined
    return t('validate.https.proxy.url.not.valid')
  }
  return undefined
}

export function validateHttpsURL(value: string, t: TFunction) {
  if (value) {
    if (
      validator.isURL(value, {
        require_protocol: true,
        require_valid_protocol: true,
        protocols: ['https'],
        require_host: true,
      })
    ) {
      return undefined
    }
    return t('validate.https.url.not.valid')
  }
  return undefined
}

const FQDN_OPTIONS = { require_tld: false }
export function validateNoProxy(value: string, t: TFunction) {
  if (
    validator.isFQDN(value, FQDN_OPTIONS) ||
    (value.startsWith('.') && validator.isFQDN(value.substring(1), FQDN_OPTIONS)) ||
    validator.isIP(value) ||
    isCidr(value) ||
    value === '*'
  ) {
    return undefined
  }

  return t(
    "Each value must be a domain name (optionally prefaced with '.' to match subdomains only), IP address, other network CIDR, or '*'."
  )
}

export function validateNoProxyList(value: string, t: TFunction) {
  if (value) {
    const noProxies = value.split(',')
    if (noProxies) {
      let validationMessage
      const validationFailed = noProxies.some((noProxy) => {
        validationMessage = validateNoProxy(noProxy, t)
        return !!validationMessage
      })
      if (validationFailed) {
        return validationMessage
      }
    }
  }
  return undefined
}

export function validateKubernetesResourceName(value: string, _item: unknown, t?: TFunction) {
  t = t ? t : (value) => value
  if (!value) return undefined
  if (value.length > 253) return `${t('This value can contain at most 253 characters')}`
  for (const char of value) {
    if (!lowercaseAlphaNumericCharacters.includes(char) && char !== '-' && char !== '.')
      return `${t("This value can only contain lowercase alphanumeric characters or '-' or '.'")}`
  }
  if (!lowercaseAlphaNumericCharacters.includes(value[0]))
    return `${t('This value must start with an alphanumeric character')}`
  if (!lowercaseAlphaNumericCharacters.includes(value[value.length - 1]))
    return `${t('This value must end with an alphanumeric character')}`
  return undefined
}

export function validatePolicyName(value: string, resource: unknown, t?: TFunction) {
  t = t ? t : (value) => value
  const error = validateKubernetesResourceName(value, t)
  if (error) return error
  const policy = resource as IResource
  const namespace = policy.metadata?.namespace ?? ''
  const combinedNameLength = value.length + namespace.length + 1

  if (combinedNameLength > 63)
    return t(
      'The combined length of namespace and policy name (namespaceName.policyName) should not exceed 63 characters'
    )
  return undefined
}

export function validateRequiredPrefix(value: string, requiredPrefix: string, t: TFunction) {
  if (value && !value?.startsWith(requiredPrefix)) {
    return t("The path must begin with '{{prefix}}'", { prefix: requiredPrefix })
  }
  return undefined
}

export function validateVCenterServer(value: string, t: TFunction) {
  if (!validator.isFQDN(value) && !validator.isIP(value)) {
    if (value && value.indexOf('://') > 0) {
      const scheme = value.split('://')[0]
      return t(
        "The value must be a fully-qualified host name or IP address. Do not include the '{{scheme}}://' URL scheme.",
        { scheme }
      )
    } else {
      return t('The value must be a fully-qualified host name or IP address.')
    }
  }
}

export function validateCidr(value: string, t: TFunction) {
  if (value == '') {
    return undefined
  }

  if (isCidr(value)) {
    return undefined
  }
  return t('Value must be a valid IPv4 CIDR.')
}
