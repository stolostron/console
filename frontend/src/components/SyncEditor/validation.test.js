/* Copyright Contributors to the Open Cluster Management project */
// import { render, waitFor } from '@testing-library/react'
// import { createBrowserHistory } from 'history'
// import { Router } from 'react-router-dom'
// import { RecoilRoot } from 'recoil'
import Ajv from 'ajv'
import { addAjvKeywords, compileAjvSchemas, validateResource } from './validation'
import schema from '../../routes/Governance/policies/schema.json'
import { keyBy } from 'lodash'

describe('validation', () => {
  it('addAjvKeywords - returns properly', async () => {
    const ajvKeywords = new Ajv({ allErrors: true, verbose: true })
    addAjvKeywords(ajvKeywords)

    expect(ajvKeywords.RULES.all.validateName).toEqual(expect.objectContaining({ keyword: 'validateName' }))
    expect(ajvKeywords.RULES.all.validateDep).toEqual(expect.objectContaining({ keyword: 'validateDep' }))
    expect(ajvKeywords.RULES.all.validateLabel).toEqual(expect.objectContaining({ keyword: 'validateLabel' }))
  })

  it('compileAJVSchemas - returns properly', async () => {
    const ajvSchemas = compileAjvSchemas(schema)

    expect(ajvSchemas).toMatchSnapshot()
  })

  it('validates incorrect compliance', async () => {
    // raw policy data for validation - ignoring lint for test readability

    // prettier-ignore
    const mappings = {"Policy":[{"apiVersion":{"$k":"apiVersion","$r":1,"$l":1,"$v":"policy.open-cluster-management.io/v1",
            "$gk":{"start":{"line":1,"col":1},"end":{"line":1,"col":11}},"$gv":{"start":{"line":1,"col":13},"end":{"line":1,"col":49}}},
            "kind":{"$k":"kind","$r":2,"$l":1,"$v":"Policy","$gk":{"start":{"line":2,"col":1},"end":{"line":2,"col":5}},
            "$gv":{"start":{"line":2,"col":7},"end":{"line":2,"col":13}}},"metadata":{"$k":"metadata","$r":3,"$l":3,
            "$v":{"name":{"$k":"name","$r":4,"$l":1,"$v":"policy-pod","$gk":{"start":{"line":4,"col":3},"end":{"line":4,"col":7}},
            "$gv":{"start":{"line":4,"col":9},"end":{"line":4,"col":19}}},"namespace":{"$k":"namespace","$r":5,"$l":1,"$v":"default",
            "$gk":{"start":{"line":5,"col":3},"end":{"line":5,"col":12}},"$gv":{"start":{"line":5,"col":14},"end":{"line":5,"col":21}}}},
            "$gk":{"start":{"line":3,"col":1},"end":{"line":3,"col":9}},"$gv":{"start":{"line":4,"col":3},"end":{"line":6,"col":1}}},
            "spec":{"$k":"spec","$r":6,"$l":8,"$v":{"disabled":{"$k":"disabled","$r":7,"$l":1,"$v":false,"$gk":{"start":{"line":7,"col":3},
            "end":{"line":7,"col":11}},"$gv":{"start":{"line":7,"col":13},"end":{"line":7,"col":18}}},"dependencies":{"$k":"dependencies",
            "$r":8,"$l":6,"$v":[{"$k":"0","$r":9,"$l":5,"$v":{"name":{"$k":"name","$r":9,"$l":1,"$v":"namespace-foo-setup-policy",
            "$gk":{"start":{"line":9,"col":7},"end":{"line":9,"col":11}},"$gv":{"start":{"line":9,"col":13},"end":{"line":9,"col":39}}},
            "namespace":{"$k":"namespace","$r":10,"$l":1,"$v":"","$gk":{"start":{"line":10,"col":7},"end":{"line":10,"col":16}},
            "$gv":{"start":{"line":10,"col":18},"end":{"line":10,"col":20}}},"apiVersion":{"$k":"apiVersion","$r":11,"$l":1,
            "$v":"policy.open-cluster-management.io/v1","$gk":{"start":{"line":11,"col":7},"end":{"line":11,"col":17}},"$gv":{"start":{"line":11,"col":19},
            "end":{"line":11,"col":55}}},"kind":{"$k":"kind","$r":12,"$l":1,"$v":"Policy","$gk":{"start":{"line":12,"col":7},"end":{"line":12,"col":11}},
            "$gv":{"start":{"line":12,"col":13},"end":{"line":12,"col":19}}},"compliance":{"$k":"compliance","$r":13,"$l":1,"$v":"Complianta",
            "$gk":{"start":{"line":13,"col":7},"end":{"line":13,"col":17}},"$gv":{"start":{"line":13,"col":19},"end":{"line":13,"col":29}}}},
            "$gv":{"start":{"line":9,"col":7},"end":{"line":14,"col":1}}}],"$gk":{"start":{"line":8,"col":3},"end":{"line":8,"col":15}},
            "$gv":{"start":{"line":9,"col":5},"end":{"line":14,"col":1}}}},"$gk":{"start":{"line":6,"col":1},"end":{"line":6,"col":5}},
            "$gv":{"start":{"line":7,"col":3},"end":{"line":14,"col":1}}}}]}

    // prettier-ignore
    const resource = {"apiVersion":"policy.open-cluster-management.io/v1","kind":"Policy","metadata":{"name":"policy-pod",
            "namespace":"default"},"spec":{"disabled":false,"dependencies":[{"name":"namespace-foo-setup-policy","namespace":"",
            "apiVersion":"policy.open-cluster-management.io/v1","kind":"Policy","compliance":"Complianta"}]}}

    const prefix = ['Policy', 0]
    const errors = []
    const ajvSchemas = compileAjvSchemas(schema)
    const validatorMap = keyBy(ajvSchemas, 'type')
    const validator = validatorMap['Policy'].validator

    validateResource(validator, mappings, prefix, resource, errors, [])

    expect(errors[0]).toEqual(
      expect.objectContaining({
        message: 'must be equal to one of the allowed values: "Compliant", "NonCompliant", "Pending"',
      })
    )
  })

  it('validates incorrect namespace in dependency', async () => {
    // raw policy data for validation - ignoring lint for test readability

    // prettier-ignore
    const mappings = {"Policy":[{"apiVersion":{"$k":"apiVersion","$r":1,"$l":1,"$v":"policy.open-cluster-management.io/v1",
            "$gk":{"start":{"line":1,"col":1},"end":{"line":1,"col":11}},"$gv":{"start":{"line":1,"col":13},"end":{"line":1,"col":49}}},
            "kind":{"$k":"kind","$r":2,"$l":1,"$v":"Policy","$gk":{"start":{"line":2,"col":1},"end":{"line":2,"col":5}},"$gv":{"start":{"line":2,"col":7},
            "end":{"line":2,"col":13}}},"metadata":{"$k":"metadata","$r":3,"$l":3,"$v":{"name":{"$k":"name","$r":4,"$l":1,"$v":"policy-pod",
            "$gk":{"start":{"line":4,"col":3},"end":{"line":4,"col":7}},"$gv":{"start":{"line":4,"col":9},"end":{"line":4,"col":19}}},
            "namespace":{"$k":"namespace","$r":5,"$l":1,"$v":"default","$gk":{"start":{"line":5,"col":3},"end":{"line":5,"col":12}},
            "$gv":{"start":{"line":5,"col":14},"end":{"line":5,"col":21}}}},"$gk":{"start":{"line":3,"col":1},"end":{"line":3,"col":9}},
            "$gv":{"start":{"line":4,"col":3},"end":{"line":6,"col":1}}},"spec":{"$k":"spec","$r":6,"$l":8,"$v":{"disabled":{"$k":"disabled",
            "$r":7,"$l":1,"$v":false,"$gk":{"start":{"line":7,"col":3},"end":{"line":7,"col":11}},"$gv":{"start":{"line":7,"col":13},
            "end":{"line":7,"col":18}}},"dependencies":{"$k":"dependencies","$r":8,"$l":6,"$v":[{"$k":"0","$r":9,"$l":5,"$v":{"name":{"$k":"name","$r":9,
            "$l":1,"$v":"namespace-foo-setup-policy","$gk":{"start":{"line":9,"col":7},"end":{"line":9,"col":11}},"$gv":{"start":{"line":9,"col":13},
            "end":{"line":9,"col":39}}},"namespace":{"$k":"namespace","$r":10,"$l":1,"$v":"default","$gk":{"start":{"line":10,"col":7},"end":{"line":10,
            "col":16}},"$gv":{"start":{"line":10,"col":18},"end":{"line":10,"col":25}}},"apiVersion":{"$k":"apiVersion","$r":11,"$l":1,
            "$v":"policy.open-cluster-management.io/v1","$gk":{"start":{"line":11,"col":7},"end":{"line":11,"col":17}},"$gv":{"start":{"line":11,"col":19},
            "end":{"line":11,"col":55}}},"kind":{"$k":"kind","$r":12,"$l":1,"$v":"ConfigurationPolicy","$gk":{"start":{"line":12,"col":7},"end":{"line":12,"col":11}},
            "$gv":{"start":{"line":12,"col":13},"end":{"line":12,"col":32}}},"compliance":{"$k":"compliance","$r":13,"$l":1,"$v":"Compliant",
            "$gk":{"start":{"line":13,"col":7},"end":{"line":13,"col":17}},"$gv":{"start":{"line":13,"col":19},"end":{"line":13,"col":28}}}},"$gv":{"start":{"line":9,
            "col":7},"end":{"line":14,"col":1}}}],"$gk":{"start":{"line":8,"col":3},"end":{"line":8,"col":15}},"$gv":{"start":{"line":9,"col":5},
            "end":{"line":14,"col":1}}}},"$gk":{"start":{"line":6,"col":1},"end":{"line":6,"col":5}},"$gv":{"start":{"line":7,"col":3},"end":{"line":14,"col":1}}}}]}

    // prettier-ignore
    const resource = {"apiVersion":"policy.open-cluster-management.io/v1","kind":"Policy","metadata":{"name":"policy-pod","namespace":"default"},
            "spec":{"disabled":false,"dependencies":[{"name":"namespace-foo-setup-policy","namespace":"default","apiVersion":"policy.open-cluster-management.io/v1",
            "kind":"ConfigurationPolicy","compliance":"Compliant"}]}}

    const prefix = ['Policy', 0]
    const errors = []
    const ajvSchemas = compileAjvSchemas(schema)
    const validatorMap = keyBy(ajvSchemas, 'type')
    const validator = validatorMap['Policy'].validator

    validateResource(validator, mappings, prefix, resource, errors, [])

    expect(errors[0]).toEqual(
      expect.objectContaining({
        message:
          'Dependencies on ConfigurationPolicies, IamPolicies, and CertificatePolicies cannot contain a namespace',
      })
    )
  })
})
