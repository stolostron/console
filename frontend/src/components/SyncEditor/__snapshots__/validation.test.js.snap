// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation compileAJVSchemas - returns properly 1`] = `
Array [
  Object {
    "required": 1,
    "type": "Policy",
    "validator": [Function],
  },
  Object {
    "required": undefined,
    "type": "Placement",
    "validator": [Function],
  },
  Object {
    "required": undefined,
    "type": "PlacementRule",
    "validator": [Function],
  },
  Object {
    "required": undefined,
    "type": "PlacementBinding",
    "validator": [Function],
  },
]
`;
